### PR TITLE
Improve TypeScript types by using variadic tuple instead of overloads

### PR DESCRIPTION
**Pull request by privatenumber**
## Problem
- Function signature was using overloads to hardcode support for up to 10 cases

## Changes
- [x] Use [variadic tuple types](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-0.html#variadic-tuple-types) to support all input lengths.

  This is a [breaking change in type](https://github.com/microsoft/TypeScript/issues/39788#issuecomment-665223322):

  Before:
  ```
  pAll<string, number>([p1, p2]);
  ```

  With variadic tuple:
  ```
  pAll<[string, number]>([p1, p2]);
  ```

- [x] Tests are passing

## References
- Variadic tuple in Promise.all issue https://github.com/microsoft/TypeScript/issues/39788
- Variadic tuple to Promise.all experiment https://github.com/microsoft/TypeScript/pull/39796

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -29,158 +32,10 @@ import got = require('got');
...
-		options?: pAll.Options
-	): Promise<[Result1]>;
-	<TAll>(
-		tasks: Iterable<pAll.PromiseFactory<TAll>> | Array<pAll.PromiseFactory<TAll>>,
-		options?: pAll.Options
-	): Promise<TAll[]>;
+	<Task extends pAll.PromiseFactory<unknown>[]>(
+		tasks: readonly [...Task],
+		options?: pAll.Options,
+	): Promise<{ [P in keyof Task]: Awaited<ReturnType<Task[P]>> }>
```
Comment (privatenumber) [inline index.d.ts:38]: Getting a TS error on the `Task[P]`:

```
Type 'Task[P]' does not satisfy the constraint '(...args: any) => any'.
  Type 'Task[keyof Task]' is not assignable to type '(...args: any) => any'.
    Type 'Task[string] | Task[number] | Task[symbol]' is not assignable to type '(...args: any) => any'.
      Type 'Task[string]' is not assignable to type '(...args: any) => any'.ts(2344)
```

I'm not sure how to resolve this. Would love any help on this.

Comment (sindresorhus) [reply index.d.ts:38]: `pAll` returns a promise for an array of fulfilled values. `{ [P in keyof Task]: Awaited<ReturnType<Task[P]>> }>` is not an array.

Comment (privatenumber) [reply index.d.ts:38]: The type tests are passing so it should be returning an array of fulfilled values. Not to mention, I think it's necessary to use the `{ [P in key of Array]: Array[P] }` syntax to map over the array.


I also asked a similar question (related but different problem) on Stackoverflow. Hopefully will find a way to resolve this soon.

Comment (privatenumber) [reply index.d.ts]: Got it working! Just needed to make sure that `Task[P]` was a function. Simple solution 🤦‍♀️

**Comment by privatenumber** [comment]
Looks like tsd was outdated and didn't support the variadic tuples syntax. Works fine now.

**Comment by privatenumber** [comment]
Looks like one of `tsd@0.11.0`'s dependencies was previously satisfying a `typescript` dependency required by `@typescript-eslint` in `xo` but is no longer satisfied:

<details>
<summary>Cannot find module 'typescript'</summary>

```
Error: Failed to load parser '/home/runner/work/p-all/p-all/node_modules/@typescript-eslint/parser/dist/parser.js' declared in 'BaseConfig': Cannot find module 'typescript'
Require stack:
- /home/runner/work/p-all/p-all/node_modules/@typescript-eslint/typescript-estree/dist/parser.js
- /home/runner/work/p-all/p-all/node_modules/@typescript-eslint/parser/dist/parser.js
- /home/runner/work/p-all/p-all/node_modules/eslint/lib/cli-engine/config-array-factory.js
- /home/runner/work/p-all/p-all/node_modules/eslint/lib/cli-engine/cascading-config-array-factory.js
- /home/runner/work/p-all/p-all/node_modules/eslint/lib/cli-engine/cli-engine.js
- /home/runner/work/p-all/p-all/node_modules/eslint/lib/cli-engine/index.js
- /home/runner/work/p-all/p-all/node_modules/eslint/lib/api.js
- /home/runner/work/p-all/p-all/node_modules/xo/index.js
- /home/runner/work/p-all/p-all/node_modules/xo/cli-main.js
- /home/runner/work/p-all/p-all/node_modules/xo/cli.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:885:15)
    at Function.Module._load (internal/modules/cjs/loader.js:730:27)
    at Module.require (internal/modules/cjs/loader.js:957:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at Object.<anonymous> (/home/runner/work/p-all/p-all/node_modules/@typescript-eslint/typescript-estree/dist/parser.js:20:25)
    at Module._compile (internal/modules/cjs/loader.js:1068:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1097:10)
    at Module.load (internal/modules/cjs/loader.js:933:32)
    at Function.Module._load (internal/modules/cjs/loader.js:774:14)
    at Module.require (internal/modules/cjs/loader.js:957:19)
npm ERR! Test failed.  See above for more details.
```

</details>

Installing the latest `typescript` as a devDependency yields:

<details>
<summary>Parsing error: Cannot read property map of undefined</summary>

```
$ npm test

> p-all@3.0.0 test /Users/osame/Documents/public-github/sindresorhus/p-all
> xo && ava && tsd

  index.test-d.ts
  ✖  Parsing error: Cannot read property map of undefined

  index.d.ts
  ✖  Parsing error: Cannot read property map of undefined

  2 errors
```

</details>

Looks like an internal error within xo, perhaps an incompatibility between `@typescript-eslint/parser` and the latest typescript. There's [no expected version of `typescript` specified in their package.json](https://unpkg.com/@typescript-eslint/parser@2.34.0/package.json).


Upgrading `xo` to the latest yields:

<details>
<summary>New linting errors</summary>

```
$ npm test

> p-all@3.0.0 test /Users/osame/Documents/public-github/sindresorhus/p-all
> xo && ava && tsd

  index.test-d.ts:2:15
  ✖   2:15   A require() style import is forbidden.                                 @typescript-eslint/no-require-imports
  ✖  16:7    Unsafe assignment of an any value.                                     @typescript-eslint/no-unsafe-assignment
  ✖  18:20   Unsafe member access [0] on an any value.                              @typescript-eslint/no-unsafe-member-access
  ✖  19:20   Unsafe member access [1] on an any value.                              @typescript-eslint/no-unsafe-member-access
  ✖  20:18   Unsafe member access [2] on an any value.                              @typescript-eslint/no-unsafe-member-access
  ✖  21:20   Unsafe member access [3] on an any value.                              @typescript-eslint/no-unsafe-member-access

  index.d.ts:34:15
  ⚠  10:1    Unexpected todo comment: TODO: Refactor the whole definition back....  no-warning-comments
  ✖  34:15   pAll is already defined.                                               @typescript-eslint/no-redeclare
  ✖  37:15   There should be no space after {.                                      @typescript-eslint/object-curly-spacing
  ✖  37:110  There should be no space before }.                                     @typescript-eslint/object-curly-spacing

  index.js:1:1
  ✖   1:1    Do not use "use strict" directive.                                     unicorn/prefer-module
  ✖   2:14   Do not use "require".                                                  unicorn/prefer-module
  ✖   4:1    Do not use "module".                                                   unicorn/prefer-module

  1 warning
  12 errors
```

</details>

I would have to add a `tsconfig.json` to eliminate the `@typescript-eslint/no-unsafe-member-access` errors. I can copy the [default config that `tsd` uses](https://github.com/SamVerschueren/tsd#custom-typescript-config), but I know you usually don't have these in your repo. I can disable the prefer-module rules for now.


How would you like me to move forward?

**Comment by sindresorhus** [comment]
I'll handle the linting stuff.

**Pull request merged**